src/transaction/gas_price.rs

```diff
@@ -0,0 +1,274 @@
+//! Implementation of gas price estimation.
+
+use crate::conv;
+use crate::future::CompatCallFuture;
+use futures::compat::Future01CompatExt;
+use futures::future::OptionFuture;
+use pin_project::{pin_project, project};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use web3::api::Web3;
+use web3::error::Error as Web3Error;
+use web3::types::U256;
+use web3::Transport;
+
+/// The gas price setting to use.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum GasPrice {
+    /// The standard estimated gas price from the node, this is usually the
+    /// median gas price from the last few blocks. This is the default gas price
+    /// used by transactions.
+    Standard,
+    /// A factor of the estimated gas price from the node. `GasPrice::Standard`
+    /// is equivalent to `GasPrice::Scaled(1.0)` (not considering possible
+    /// rounding errors).
+    Scaled(f64),
+    /// Specify a specific gas price to use for the transaction. This will cause
+    /// the transaction `SendFuture` to not query the node for a gas price
+    /// estimation.
+    Value(U256),
+}
+
+impl GasPrice {
+    /// A low gas price. Using this may result in long confirmation times for
+    /// transactions, or the transactions not being mined at all.
+    pub fn low() -> Self {
+        GasPrice::Scaled(0.8)
+    }
+
+    /// A high gas price that usually results in faster mining times.
+    /// transactions, or the transactions not being mined at all.
+    pub fn high() -> Self {
+        GasPrice::Scaled(6.0)
+    }
+
+    /// Returns `Some(value)` if the gas price is explicitly specified, `None`
+    /// otherwise.
+    pub fn value(&self) -> Option<U256> {
+        match self {
+            GasPrice::Value(value) => Some(*value),
+            _ => None,
+        }
+    }
+
+    /// Resolves the gas price into a value. Returns a future that resolves once
+    /// the gas price is calculated as this may require contacting the node for
+    /// gas price estimates in the case of `GasPrice::Standard` and
+    /// `GasPrice::Scaled`.
+    pub fn resolve<T: Transport>(web3: &Web3<T>, gas_price: Self) -> ResolveGasPriceFuture<T> {
```

Comment:
This method makes testing easier? Or why don't we want to inline it?

Reply:
Do you mean why not make it take `&self`? For consistency with `GasPrice::resolve_for_transaction_request`. Not really attached to it being an associated (static) function and if you find it easier to read if it took a `self` parameter I can change it.

Reply:
The other motivating factor was that conceptually, I felt like `Web3::resolve_gas_price(GasPrice)` makes more to me, in that it should be a method on `Web3` and not `GasPrice`.

Reply:
I changed it :)